### PR TITLE
Add multi_gru_fuse_pass and tests

### PR DESCRIPTION
**Pull request by wojtuss**
### PR types
New features

### PR changes
Others

### Describe
This PR introduces a `multi_gru_fuse_pass` which fuses two concatenated `fusion_gru` ops into a single layer `multi_gru` op.
This fuse is the first step in transformation of a sequence of concatenated pairs of `fusion_gru` ops into one multilayer `multi_gru` op.

This PR is a part of the draft PR #28230.
It is related to the GRU INT8 model optimization (#27330).

### CODE REVIEW

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc [hunk 1]
```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wozna) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: 2019 should be changed to 2020

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wozna) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h]: Ditto 2020

**Review by wozna — DISMISSED** [review]
Only cosmetic changes. All LGTM

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc (hunk 1 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: Done.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h (hunk 2 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h]: Done.

**Review by wozna — DISMISSED** [review]
LGTM

**Comment by wojtuss** [comment]
@luotao1 , could you please add your review?

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+namespace paddle {
+namespace framework {
+namespace ir {
+
+using EigenVectorArrayMap = Eigen::Map<Eigen::Array<double, Eigen::Dynamic, 1>>;
+using string::PrettyLogDetail;
+
+namespace {
+
+void UnlinkNodes(ir::Node* a, ir::Node* b) {
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: This function is unused.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  a->outputs.erase(std::remove(a->outputs.begin(), a->outputs.end(), b),
+                   a->outputs.end());
+  b->inputs.erase(std::remove(b->inputs.begin(), b->inputs.end(), a),
+                  b->inputs.end());
+}
+
+std::vector<std::string> join_inputs(Node* op1, Node* op2,
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: Don't we have camel case convention for naming functions?

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+#include "paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h"
+#include <limits>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/ir/graph_pattern_detector.h"
+#include "paddle/fluid/platform/errors.h"
+#include "paddle/fluid/platform/mkldnn_helper.h"
+#include "paddle/fluid/string/pretty_log.h"
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]:
```suggestion
#include <algorithm>
#include <vector>
#include "paddle/fluid/framework/eigen.h"
#include "paddle/fluid/framework/ir/graph_pattern_detector.h"
#include "paddle/fluid/platform/errors.h"
#include "paddle/fluid/string/pretty_log.h"
```

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include "paddle/fluid/framework/ir/fuse_pass_base.h"
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/graph_pattern_detector.h"
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h]:
```suggestion
#include <string>

#include "paddle/fluid/framework/ir/fuse_pass_base.h"
#include "paddle/fluid/framework/ir/graph.h"
```

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+class MultiGRUFusePass : public FusePassBase {
```
Comment (arogowie-intel) [inline paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h:31]: Please add some doc on what pattern this fuse works on.

**Review by arlesniak — COMMENTED** [review]
Thank you for great job Wojtek! 
Just a question, is other pattern order than: fusion_gru(is_reverse=false) is the first input to concat and fusion_gru(is_reverse=true) the second one, to be concerned in some cases?

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc (hunk 3 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: Done.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc (hunk 4 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: Done.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc (hunk 5 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.cc]: Done.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h (hunk 6 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h]: Done.

File: paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h (hunk 7 above)
Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/multi_gru_fuse_pass.h:31]: Done.

**Comment by wojtuss** [comment]
> Thank you for great job Wojtek!
> Just a question, is other pattern order than: fusion_gru(is_reverse=false) is the first input to concat and fusion_gru(is_reverse=true) the second one, to be concerned in some cases?

@arlesniak,
You are right. Now it is fixed. The inputs to concat can come in any order.

**Review by arogowie-intel — APPROVED** [review]

**Review by arlesniak — APPROVED** [review]
LGTM

**Comment by wojtuss** [comment]
@luotao1 , could you please add your review?

**Review by luotao1 — APPROVED** [review]

**Pull request merged**